lib/events/pgevents/pgevents.go

```diff
@@ -0,0 +1,593 @@
+		}
+		cfg.PoolConfig.BeforeConnect = bc
+	}
+
+	cfg.Log.Info("Setting up events backend.")
+
+	pgcommon.TryEnsureDatabase(ctx, cfg.PoolConfig, cfg.Log)
+
+	pool, err := pgxpool.NewWithConfig(ctx, cfg.PoolConfig)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	periodicCtx, cancel := context.WithCancel(context.Background())
+	l := &Log{
+		log:    cfg.Log,
+		pool:   pool,
+		cancel: cancel,
+	}
+
+	if err := l.setupAndMigrate(ctx); err != nil {
+		l.Close()
+		return nil, trace.Wrap(err)
+	}
+
+	if !cfg.DisableCleanup {
+		l.wg.Add(1)
+		go l.periodicCleanup(periodicCtx, cfg.CleanupInterval, cfg.RetentionPeriod)
+	}
+
+	l.log.Info("Started events backend.")
+
+	return l, nil
+}
+
+// Log is an external [events.AuditLogger] backed by a PostgreSQL database.
+type Log struct {
+	log  logrus.FieldLogger
+	pool *pgxpool.Pool
+
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+}
+
+// Close stops all background tasks and closes the connection pool.
+func (l *Log) Close() error {
+	l.cancel()
+	l.wg.Wait()
+	l.pool.Close()
+	return nil
+}
+
+var schemas = []string{
+	`CREATE TABLE events (
+		event_time timestamptz NOT NULL,
+		event_id uuid NOT NULL,
+		event_type text NOT NULL,
+		session_id uuid NOT NULL,
+		event_data json NOT NULL,
+		creation_time timestamptz NOT NULL DEFAULT now(),
+		CONSTRAINT events_pkey PRIMARY KEY (event_time, event_id)
+	);
+	CREATE INDEX events_creation_time_idx ON events USING brin (creation_time);
+	CREATE INDEX events_search_session_events_idx ON events (session_id, event_time, event_id)
+		WHERE session_id != '00000000-0000-0000-0000-000000000000';`,
+}
+
+// setupAndMigrate sets up the database schema, applying the migrations in the
+// [schemas] slice starting from the first non-applied one.
+func (l *Log) setupAndMigrate(ctx context.Context) error {
+	var version int32
+	var migrateErr error
+
+	// this is split off from the rest because we might not have permissions to
+	// CREATE TABLE, which is checked even if the table exists
+	if _, err := pgcommon.RetryIdempotent(ctx, l.log, func() (struct{}, error) {
+		_, err := l.pool.Exec(ctx,
+			`CREATE TABLE IF NOT EXISTS audit_version (
+				version integer PRIMARY KEY CHECK (version > 0),
+				created timestamptz NOT NULL DEFAULT now()
+			)`, pgx.QueryExecModeExec,
+		)
+		return struct{}{}, trace.Wrap(err)
+	}); err != nil {
+		// the very first SELECT in the next transaction will fail, we don't
+		// need anything higher than debug here
+		l.log.WithError(err).Debug("Failed to confirm the existence of the audit_version table.")
+	}
+
+	const idempotent = true
+	if err := pgcommon.RetryTx(ctx, l.log, l.pool, pgx.TxOptions{
+		IsoLevel:   pgx.Serializable,
+		AccessMode: pgx.ReadWrite,
+	}, idempotent, func(tx pgx.Tx) error {
+		if err := tx.QueryRow(ctx,
+			"SELECT COALESCE(max(version), 0) FROM audit_version",
+			pgx.QueryExecModeExec,
+		).Scan(&version); err != nil {
+			return trace.Wrap(err)
+		}
+
+		if int(version) > len(schemas) {
```

Comment:
From what I understand, we will scale auth to 1, run this and scale back? That also means that rollback after won't be possible (when the cheam changes)?

Reply:
I'm not really expecting us to do changes to the schema, but yes, a low-level schema change will not be rollbackable without some manual intervention on the database. It should be fine to start multiple new auths all at the same version however, the setup transaction will just be retried I reckon.